pkg/migration-tool/writer/writer.go

```diff
@@ -0,0 +1,153 @@
+package writer
+
+import (
+	"context"
+	"fmt"
+	"go.uber.org/atomic"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/migration-tool/planner"
+)
+
+const (
+	DefaultWriteTimeout  = time.Minute * 5
+	BackOffRetryDuration = time.Millisecond * 100
+)
+
+type RemoteWrite struct {
+	sigBlockRead  chan struct{}
+	sigBlockWrite chan struct{} // To the reader.
+	plan          *planner.Plan
+	client        remote.WriteClient
+}
+
+// New returns a new remote write. It is responsible for writing to the remote write storage.
+func New(remoteWriteUrl string, plan *planner.Plan, sigRead, sigWrite chan struct{}) (*RemoteWrite, error) {
+	parsedUrl, err := url.Parse(remoteWriteUrl)
+	if err != nil {
+		return nil, fmt.Errorf("url-parse: %w", err)
+	}
+	wc, err := remote.NewWriteClient(fmt.Sprintf("writer-%d", 1), &remote.ClientConfig{
+		URL:     &config.URL{URL: parsedUrl},
+		Timeout: model.Duration(DefaultWriteTimeout),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating write-client: %w", err)
+	}
+	write := &RemoteWrite{
+		plan:          plan,
+		client:        wc,
+		sigBlockRead:  sigRead,
+		sigBlockWrite: sigWrite,
+	}
+	return write, nil
+}
+
+// Run runs the remote-writer. It waits for the remote-reader to give access to the in-memory
+// data-block that is written after the most recent fetch. After reading the block and storing
+// the data locally, it gives back the writing access to the remote-reader for further fetches.
+func (rw *RemoteWrite) Run(wg *sync.WaitGroup, readerUp *atomic.Bool) error {
+	log.Info("msg", "writer is up")
+	defer func() {
+		wg.Done()
+	}()
+	for {
+		if _, ok := <-rw.sigBlockRead; !ok {
+			break
+		}
+		var buf []byte
+		blockRef := rw.plan.CurrentBlock()
+		blockRef.SetDescription("pushing...", 1)
+		ts := timeseriesRefToTimeseries(blockRef.Timeseries)
+		if err := rw.sendSamplesWithBackoff(context.Background(), ts, &buf); err != nil {
+			return fmt.Errorf("remote-write run: %w", err)
+		}
+		if err := blockRef.Done(); err != nil {
+			return fmt.Errorf("remote-write: %w", err)
+		}
+		rw.sigBlockWrite <- struct{}{}
+		if !readerUp.Load() {
```

Comment:
can the sigBlockRead channel be used to signal whether the reader is up or not? it will close upon existing the Reader.Run function

Reply:
Yeah, that was the plan during refactoring. Done now.